Documentation: add migration document

squeed opened the pull request:
This document explains the basics of how to migrate Cilium on a running cluster. It includes a copy-and-paste example for migrating off of Flannel.

It utilizes the per-node-configuration override feature to gradually roll out Cilium. It configures a hybrid cluster with multiple overlays so that all pods have full connectivity until migration is complete.

qmonnet commented on Documentation/installation/k8s-install-migration.rst:
```diff
@@ -0,0 +1,310 @@
+
+This document outlines how migrations with Cilium work. You will have a good
+understanding of the basic requirements, as well as see an example migration
+which you can practice using :ref:`Kind <gs_kind>`
```
```suggestion
which you can practice using :ref:`Kind <gs_kind>`.
```

qmonnet commented on Documentation/installation/k8s-install-migration.rst:
```diff
@@ -0,0 +1,310 @@
+be re-enabled.
+
+It is strongly recommended to install Cilium using the :ref:`cluster-pool <ipam_crd_cluster_pool>`
+IPAM allocator.  This provides the strongest assurance that there will
```
```suggestion
IPAM allocator. This provides the strongest assurance that there will
```

qmonnet commented on Documentation/installation/k8s-install-migration.rst:
```diff
@@ -0,0 +1,310 @@
+
+In summary, the process looks like:
+
+1. Install cilium in "secondary" mode.
```
```suggestion
1. Install cilium in "secondary" mode
```

qmonnet commented on Documentation/installation/k8s-install-migration.rst:
```diff
@@ -0,0 +1,310 @@
+Preparation
+-----------
+
+Optional: Create a :ref:`Kind <gs_kind>` cluster and install Flannel on it
```
```suggestion
Optional: Create a :ref:`Kind <gs_kind>` cluster and install Flannel on it.
```

qmonnet commented on Documentation/installation/k8s-install-migration.rst:
```diff
@@ -0,0 +1,310 @@
+      disableDefaultCNI: true
+    EOF
+    $ kind create cluster --config=kind-config.yaml
+    $ kubectl apply -n kube-system --server-side -f \ |SCM_WEB|\/examples/misc/migration/install-reference-cni-plugins.yaml
```
You need this block to be a `.. parsed-literal::`, or the substitution for `|SCM_WEB|` will not happen.

qmonnet commented on Documentation/installation/k8s-install-migration.rst:
```diff
@@ -0,0 +1,310 @@
+        ipam:
+          mode: "cluster-pool"
+          operator:
+            clusterPoolIPv4PodCIDRList: ["10.245.0.0/16"] # Migration: ensure this is distinct and unused
```
```suggestion
        tunnel: vxlan # Migration: Optional: default is GENEVE, configure as needed
        tunnelPort: 8473 # Migration: Optional, change only if both networks use the same port by default
        cni:
          customConf: true # Migration: Don't install a CNI configuration file
        ipam:
          mode: "cluster-pool"
          operator:
            clusterPoolIPv4PodCIDRList: ["10.245.0.0/16"] # Migration: Ensure this is distinct and unused
```

qmonnet commented on Documentation/installation/k8s-install-migration.rst:
```diff
@@ -0,0 +1,310 @@
+
+.. code-block:: shell-session
+
+  export NODE="kind-worker" # for the kind example
```
No need to `export`, I think it's not necessary here (plus the `export` syntax varies between shells).

```suggestion
  $ NODE="kind-worker" # for the Kind example
```

qmonnet commented on Documentation/installation/k8s-install-migration.rst:
```diff
@@ -0,0 +1,310 @@
+    **Cons:**
+
+    - Pods may be restarted twice during the migration
+    - Any potential issues may be noticed sooner
```
Is this one a disadvantage?

squeed replied:
The wording is funny, you're right. I meant to say that potential issues are more likely to show up if you don't drain. I'll re-phrase.

qmonnet commented on Documentation/installation/k8s-install-migration.rst:
```diff
@@ -0,0 +1,310 @@
+      $ kubectl cordon $NODE
+      $ kubectl drain $NODE
+
+2.  Label the node. This causes the CiliumNodeConfig to apply to this node.
```
```suggestion
2.  Label the node. This causes the ``CiliumNodeConfig`` to apply to this node.
```

qmonnet commented on Documentation/installation/k8s-install-migration.rst:
```diff
@@ -0,0 +1,310 @@
+
+    .. code-block:: shell-session
+
+      kubectl -n kube-system delete pod --field-selector spec.nodeName=$NODE -l k8s-app=cilium
```
```suggestion
    If using Kind, simulate a reboot by restarting Cilium on the node:

    .. code-block:: shell-session

      $ kubectl -n kube-system delete pod --field-selector spec.nodeName=$NODE -l k8s-app=cilium
```

mauilion replied:
you can simulate a node reboot with docker restart `kind-node`

mauilion replied:
and you should cause this clears out old not in use interfaces and erroneuos iptables things

qmonnet commented on examples/misc/migration/install-reference-cni-plugins.yaml:
```diff
@@ -0,0 +1,63 @@
+      volumes:
+      - name: cni-plugin
+        hostPath:
+          path: /opt/cni/bin
```
Nit: No `\n` at end of file.

![image](https://user-images.githubusercontent.com/17001771/219021549-9300d975-1683-4d25-b59c-596255c8d267.png)

qmonnet reviewed (changes requested):
Contents seem all good to me, I just have a few minor comments on formatting (plus nits on case and punctuation, for consistency).

mauilion commented on Documentation/installation/k8s-install-migration.rst:
```diff
@@ -0,0 +1,310 @@
+
+Cilium supports "adopting" existing clusters. Running clusters can
+be migrated on a node-by-node basis, without disrupting existing traffic
+or requiring a complete cluster outage or rebuild.
```
I
```suggestion
Cilium can be used to migrate from another cni. Running clusters can
be migrated on a node-by-node basis, without disrupting existing traffic
or requiring a complete cluster outage or rebuild depending on the complexity of the migration case.
```

mauilion commented on Documentation/installation/k8s-install-migration.rst:
```diff
@@ -0,0 +1,310 @@
+across the cluster. While pods on a given node can only be attached to one network,
+they have access to both Cilium and non-Cilium pods while the migration is
+taking place. As long as Cilium and the existing networking provider use a separate
+IP range, the Linux routing table takes care of separating traffic.
```
```suggestion
When the kubelet creates a Pod, the installed CNI, as configured in /etc/cni/net.d/, is called. The cni will handle the networking for a pod. From allocating an ip address to ensuring that the pod interface is configured appriopriately as part of the pod network. The Pods network configuration shares the same life cycle
as the Pod, In the case of migration, we typically reconfigure /etc/cni/net.d/ to point to the new cni. Once completed any existing pods will have been configured by the old cni and any new pods will be configured by the newer
CNI. To complete the migration all Pods on the cluster that are
configured by the old cni must be recycled in order to be a member of the new CNI.

A naive approach to migrating a CNI would be to reconfigure all nodes with a new CNI and then gradually restart each node in the
cluster, thus replacing the CNI when the node is brought back up and ensuring that all pods are part of the new CNI.

This simple migration while effective comes at the cost of restarting all pods and nodes cluster wide before proceeding. This will be a significant outage for the cluster.

In this document we will discuss a model for live migrating between two deployed CNI implementations. This will have the benefit of reducing downtime of nodes and workloads and ensuring that workloads on both configured CNIs can communicate during migration. 

For live migration to work the second CNI will be installed with a separate
CIDR range to that of the currently installed CNI, which has no knowledge of
the other network range. Not to mention the use of different encapsulation
protocols. If the cluster is making use of network policies or other capabilities that are cni specific. This usually adds significant challenges to the migration model.

To help address the case of live migration as of 1.13 Cilium supports a "hybrid" mode.

This migration strategy allows us to configure each nodes CNI implementation independently. Allowing you to migrate nodes at your own pace and ensuring that workloads on those nodes are able to reach other workloads still deployed using the old CNI while the migration is taking place. 
 As long as Cilium and the existing CNI use a separate
IP range, the Linux routing table takes care of separating traffic.
```

mauilion reviewed (changes requested):
Overall I think it's fine for us to include this in the documentation. I definitely think that describing this process as simple would be a mistake. There will always be edge cases that make this difficult to manage.

nvibert commented on Documentation/installation/k8s-install-migration.rst:
```diff
@@ -0,0 +1,310 @@
+    .. code-block:: shell-session
+
+      $ kubectl cordon $NODE
+      $ kubectl drain $NODE
```
When going through the tutorial, when I drain, I get some error messages draining the node from the Flannel and Cilium Daemonset.

Should we be draining the Node using `--ignore-daemonsets` ?

```
There are pending nodes to be drained:
 kind-worker
cannot delete DaemonSet-managed Pods (use --ignore-daemonsets to ignore): kube-flannel/kube-flannel-ds-rf9b4, kube-system/cilium-2j2lb, kube-system/install-cni-plugins-md4gh, kube-system/kube-proxy-zhxcg
```

squeed replied:
Yup, good catch.

nvibert commented on Documentation/installation/k8s-install-migration.rst:
```diff
@@ -0,0 +1,310 @@
+        bpf:
+          hostLegacyRouting: true # Migration: Allow for routing between Cilium and the existing overlay
+
+4.  Optional: auto-detect settings using the :ref:`cilium-cli <install_cilium_cli>`.
```
I don't think this should be "Optional" as if a reader decided to skip through the steps, they'd miss this step and the creation of the values-initial.yaml file which is required for the following step.

squeed replied:
fair enough; updated

qmonnet reviewed (approved):
Looks good from my side. Thanks!

qmonnet commented on Documentation/installation/k8s-install-migration.rst:
```diff
@@ -0,0 +1,332 @@
+When the kubelet creates a Pod's Sandbox, the installed CNI, as configured in ``/etc/cni/net.d/``,
+is called. The cni will handle the networking for a pod - including allocating 
+an ip address, creating & configuring a network interface, and (potentially)
+establishing an overlay network. The Pod'ss network configuration shares the
```
Oops, the spell checker found a typo

```suggestion
establishing an overlay network. The Pod's network configuration shares the
```

squeed replied:
oh no, i totally meant that :-p

nvibert commented on Documentation/installation/k8s-install-migration.rst:
```diff
@@ -0,0 +1,340 @@
+        operator:
+          unmanagedPodWatcher:
+            restart: false # Migration: Don't restart unmigrated pods
+        tunnel: vxlan # Migration: Optional: default is GENEVE, configure as needed
```
Is Geneve definitely the default? Helm default value is VXLAN:
https://github.com/cilium/cilium/blob/c9723a8df3cfa336da1f8457a864105d8349acfe/install/kubernetes/cilium/values.yaml#L1681

squeed replied:
Oh, you're totally right. Whoops.

mauilion commented on Documentation/installation/k8s-install-migration.rst:
```diff
@@ -0,0 +1,351 @@
+        --overrides='{"spec": {"nodeName": "'$NODE'", "tolerations": [{"operator": "Exists"}]}}' \
+        --image ghcr.io/nicolaka/netshoot:v0.8 -- /bin/bash -c 'ip -br addr && curl -s -k https://$KUBERNETES_SERVICE_HOST/healthz && echo'
+
+    Ensure the IP address of the pod is in the Cilium CIDR(s) supplied above.
```
Ensure the IP address of the pod is in the Cilium CIDR(s) supplied above and that the apiservice is reachable.

mauilion reviewed (approved):
looks great with a minor nit.

squeed committed:
Documentation: add migration document

This document explains the basics of how to migrate Cilium on a running
cluster. It includes a copy-and-paste example for migrating off of
Flannel.

It utilizes the per-node-configuration override feature to gradually
roll out Cilium. It configures a hybrid cluster with multiple overlays
so that all pods have full connectivity until migration is complete.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>
Co-authored-by: Duffie Cooley <dcooley@isovalent.com>
Co-authored-by: Quentin Monnet <qmonnet+github@qoba.lt>
